nats_test.go

```diff
@@ -2540,3 +2541,76 @@ func TestLameDuckMode(t *testing.T) {
 	}
 	wg.Wait()
 }
+
+func TestJSSubscribe(t *testing.T) {
+	srv, _, _, nc := startJetStream(t)
+	defer os.RemoveAll(srv.JetStreamConfig().StoreDir)
+	defer srv.Shutdown()
+	defer nc.Close()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 250*time.Millisecond)
+	defer cancel()
+
+	seen := 0
+	cb := func(m *Msg) {
+		m.Ack()
+		seen++
+		if seen == 20 {
+			cancel()
+		}
+	}
+
+	cfg := &ConsumerConfig{
+		Durable:        "nats",
+		DeliverSubject: NewInbox(),
```

Comment:
I feel we should hide this by default and handle cleanly. We can expose if they really want it but by default when we subscribe should just do the right thing. We do not need to know about the inbox etc.

Reply:
this only makes sense as long as we only support 1 client per push consumer, if someone wanted another we need to give them the option to set it, so i could if say sub=="" AND DeliverySubject=="" then make an Inbox but leave it as is if a subject is given.

Reply:
We get sub on normal Subscribe() call, but yes if we are doing push based subscriber vs SubscribeSync() then spin up an inbox delivery subject if its not present.

Reply:
As things stand `SubscribeSync` is also a push under the covers for the moment, so will handle the case where no subject is given at all by setting inbox.

Reply:
The problem with this is is subj on Subscribe is empty and the DeliverySubject is empty - we dont know the users intention, they could be trying to make a pull consumer.  We abuse DeliverySubject to convey the intent to make a push/pull.

So I cant use a Inbox if `subj==""&&DeliverySubject==""` because the user might be trying to make a pull consumer but mistakenly giving it to a normal subscribe, so for now we fail in that case.  Later when we have a way to do fast pull consumers here we can revisit this perhaps?

Reply:
I think we do pull for SubscribeSync..

Reply:
I can drop the subscribe options till we have pull support from SubscribeSync() or leave as is and it works at least? We dont have pull support so not sure what the action you want here

Reply:
Subscribe should always be push based, so if they do not supply deliver subject in config create inbox etc.

Reply:
I dont think a API should change the users intent ever, if the user asks the impossible tell him so is a much better behavour, users could be pre-creating a pull consumer and try to use it via Subscribe, thats an error that we can give a nice friendly error for rather than the opaque "consumer already exist" he'd get from the server

Reply:
Not following. I am just saying that if a consumer cfg option is passed to subscribe without a delivery subject create one and do not force the user to do all the boilerplate. If it is set use it as is..

Reply:
Without a delivery subject == pull consumer.  How do I know he isnt intending to access a pull consumer? Do I just override his wish and make a pull one, or educate him about this mistake. 

For queue and sync based subscribes it does it for the user

Reply:
Subscribe forces user to provide a callback, that has always been pushed based. IMO, pull based for JetStream is equivalent to SubscribeSync where the user does NextMsg().

Reply:
I understand that! but people make mistakes.  Whatever this is not helpful, will put in the behaviour you want.

Reply:
They dont have to make a low level subscribe. This code does it for them.  We're arguing over one line that would make the behaviour clear and inform a user when their intent doesnt match what the underlying code can implement for them.

Reply:
I think if its empty create one and do low level subscribe etc. If deliver subject is set, do nothing..

What are you exactly recommending? Apologies, 50 balls in the air already this am.